pgjdbc/src/main/java/org/postgresql/jdbc/EscapedFunctions.java

```diff
@@ -683,64 +658,86 @@ private static String constantToDatePart(String type) throws SQLException {
+    zeroArgumentFunctionCall(buf, "user", "user", parsedArgs);
+  }
+
+  private static void zeroArgumentFunctionCall(StringBuilder buf, String call, String functionName,
+      List<CharSequence> parsedArgs) throws PSQLException {
     if (!parsedArgs.isEmpty()) {
-      throw new PSQLException(GT.tr("{0} function doesn''t take any argument.", "user"),
+      throw new PSQLException(GT.tr("{0} function doesn''t take any argument.", functionName),
           PSQLState.SYNTAX_ERROR);
     }
-    return "user";
+    buf.append(call);
   }
 
-  private static String singleArgumentFunctionCall(String call, String functionName,
-      List<?> parsedArgs) throws PSQLException {
+  private static void singleArgumentFunctionCall(StringBuilder buf, String call, String functionName,
+      List<CharSequence> parsedArgs) throws PSQLException {
     if (parsedArgs.size() != 1) {
       throw new PSQLException(GT.tr("{0} function takes one and only one argument.", functionName),
           PSQLState.SYNTAX_ERROR);
     }
-    StringBuilder buf = new StringBuilder();
-    buf.append(call);
-    buf.append(parsedArgs.get(0));
-    return buf.append(')').toString();
+    CharSequence arg0 = parsedArgs.get(0);
+    buf.ensureCapacity(buf.length() + call.length() + arg0.length() + 1);
+    buf.append(call).append(arg0).append(')');
   }
 
-  private static String twoArgumentsFunctionCall(String call, String functionName,
-      List<?> parsedArgs) throws PSQLException {
+  private static void twoArgumentsFunctionCall(StringBuilder buf, String call, String functionName,
+      List<CharSequence> parsedArgs) throws PSQLException {
     if (parsedArgs.size() != 2) {
       throw new PSQLException(GT.tr("{0} function takes two and only two arguments.", functionName),
           PSQLState.SYNTAX_ERROR);
     }
-    StringBuilder buf = new StringBuilder();
-    buf.append(call);
-    buf.append(parsedArgs.get(0)).append(',').append(parsedArgs.get(1));
-    return buf.append(')').toString();
+    appendCall(buf, call, ",", ")", parsedArgs);
+  }
+
+  /**
+   * Appends {@code begin arg0 separator arg1 separator end} sequence to the input {@link StringBuilder}
+   * @param sb destination StringBuilder
+   * @param begin begin string
+   * @param separator separator string
+   * @param end end string
+   * @param args arguments
+   */
+  public static void appendCall(StringBuilder sb, String begin, String separator,
+      String end, List<CharSequence> args) {
+    int size = begin.length();
+    // Avoid Iterator instantiations just in case, so plain for, not forach
+    for (int i = 0; i < args.size(); i++) {
+      size += args.get(i).length();
+    }
+    size += separator.length() * (args.size() - 1);
+    sb.ensureCapacity(sb.length() + size + 1);
+    sb.append(begin);
+    // Avoid Iterator instantiations just in case, so plain for, not forach
+    for (int i = 0; i < args.size(); i++) {
```

Comment:
don't call args.size() each time through the loop:
`for (int i=0, j=args.size(); i<j; i++)`

Reply:
Does it really matter? I find it would make code bigger => harder to reason about

Note: the code is supposed to be "internal" and it is supposed to work with `ArrayList` only.

Reply:
It looks like this no longer matters[1]. It used to be (java 1.4 and older) a performance penalty.

[1] - https://stackoverflow.com/questions/4438710/using-collection-size-in-for-loop-comparision

Reply:
I mean something like https://bugs.openjdk.java.net/browse/JDK-8166840 and/or http://2016.jpoint.ru/talks/cheremin/ (slides 69..74)